src/biguint.rs

```diff
@@ -1749,6 +1749,50 @@ impl BigUint {
         }
         acc
     }
+
+    /// Finds square root of `self`.
+    ///
+    /// The result is the greatest integer less than or equal to the
+    /// square root of `self`.
+    pub fn sqrt(&self) -> Self {
+        let one = BigUint::one();
+        let two = BigUint::from(2 as u8);
+
+        // Trivial cases
+        if self.is_zero() {
+            return BigUint::zero();
+        }
+
+        if self.is_one() {
+            return one;
+        }
+
+        // Newton's method to compute the square root of an integer.
+        //
+        // Reference:
+        // Brent & Zimmermann, Modern Computer Arithmetic, v0.5.9, Algorithm 1.13
+        //
+        // Set initial guess to something >= floor(sqrt(self)), but as low
+        // as possible to speed up convergence.
+        let bit_len = self.len() * big_digit::BITS;
+        let guess = one.shl(bit_len/2 + 1);
+
+        let mut u = guess;
+        let mut s: BigUint;
+
+        loop {
+            s = u;
+            let q = self.div_floor(&s);
+            let t: BigUint = &s + &q;
+
+            // Compute the candidate value for next iteration
+            u = t.div_floor(&two);
```

Comment:
Any reason you're using `div_floor` instead of the `/` operator?  With the operator you can also use integer literals, like `t / 2u32`.  But even more, `t >> 1` is probably faster.

Reply:
No reason whatsoever, it just slipped my mind it seems. I replaced the usage of `div_floor` with the `/` and `>>` operators as appropriate, and also replaced the `lsh` call with `<<`. Looks much cleaner now.